docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -0,0 +1,110 @@
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
+
+The current multisig public keys use amino serialization to generate the address. We will retain
+those public keys and their address formatting, and call them "legacy amino" multisig public keys
+in protobuf. We will also create multisig public keys without amino addresses to be described below.
+
+### Canonical Public Key Addresses
+
+Following on the discussion in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we propose the
+following approach.
+
+All public key types will have a unique protobuf message type such as:
+
+```proto
+package cosmos.crypto.sr25519;
+
+message PubKey {
+  bytes key = 1;
+}
+```
+ 
+All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
+These names are derived directly from .proto files in a standardized way and used
+in other places such as the type URL in `Any`s. Since there is an easy and obvious
+way to get this name for every protobuf type, we can use this message name as the
+key type prefix when creating addresses.
+
+We define the canonical address format for new (non-legacy) public keys as
+`Sha256(fmt.Sprintf("%s/%x, proto.MessageName(key), key.Bytes())[:20]`. This takes
+the first 20 bytes of an SHA-256 hash of a string with the proto message name for the key
+type joined by an `/` with the hex encoding of the key bytes.
+
+For all basic public keys, key bytes should just be the raw unencoded public key bytes.
+
+### Multisig Addresses
+
+For new multisig public keys, we define a custom address format not based on any encoding scheme
+(amino or protobuf).
+
+First we define a proto message for multisig public keys:
+```proto
+package cosmos.crypto.multisig;
+
+message PubKey {
+  uint32 threshold = 1;
+  repeated google.protobuf.Any public_keys = 2;
+}
+```
+
+Each nested public key has its own address, so we can use that address as a starting
+point for forming the multisig address. Let's create an array of strings, `hexAddresses []string`,
+which is the hex-encoded address of each nested pubkey. We join these hex encoded addresses
+with a `/`, i.e. `joinedHexAddresses := strings.Join(hexAddresses, "/")`. We then form the address of the multisig pubkey,
+using `Sha256(fmt.Sprintf("cosmos.crypto.multisig.PubKey/%d/%s", pk.Threshold, joinedHexAddresses)[:20]`.
```

Comment:
Because we have to deal with non-deterministic encoding of protobuf (i.e. ADR 027). This avoids that. Can you see if what I have their now which does sorting is okay?